fprime-util generate --ninja silences FPP error messages

| | |
|:---|:---|
|**_F´ Version_**| v3.4.3-117-g65b9e4d9d |
|**_Affected Component_**| build system |
---
## Problem Description

When I have a typo in one of my `.fpp` files, I normally get an error message explaining the problem. However, if I use `fprime-util generate --ninja` instead of `fprime-util generate`, key pieces of the error message are dropped.

```
[root@MT-400506 scythe]# fprime-util generate --ninja
[INFO] Generating build directory at: [...]
[INFO] Using toolchain file [...]/cmake/toolchain/[...].cmake for platform default
-- The C compiler identification is Clang 18.1.3
-- The CXX compiler identification is Clang 18.1.3
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [python3] python3 found at: /usr/bin/python3
-- [fpp-tools] fpp-depend found at: /usr/local/bin/fpp-depend
-- [fpp-tools] fpp-to-xml found at: /usr/local/bin/fpp-to-xml
-- [fpp-tools] fpp-to-cpp found at: /usr/local/bin/fpp-to-cpp
-- [fpp-tools] fpp-to-dict found at: /usr/local/bin/fpp-to-dict
-- [fpp-tools] fpp-locate-defs found at: /usr/local/bin/fpp-locate-defs
-- [FPRIME] Module locations: [...]
-- [FPRIME] Configuration module: [...]
-- [FPRIME] Installation directory: [...]
-- Target build toolchain/platform: [...]
-- Including [...]/cmake/platform/[...].cmake
-- [sub-build] Performing sub-build: info-cache
-- [sub-build] Generating: info-cache with target/fpp_locs;target/fpp_depend
-- [sub-build] Executing: info-cache with fpp_locs
CMake Error at tools/fprime/cmake/utilities.cmake:586 (message):
-- Configuring incomplete, errors occurred!
  [sub-build] Failed to execute: info-cache/fpp_locs:

Call Stack (most recent call first):
  tools/fprime/cmake/sub-build/sub-build.cmake:65 (execute_process_or_fail)
  tools/fprime/cmake/FPrime.cmake:162 (run_sub_build)
  tools/fprime/cmake/FPrime.cmake:231 (fprime_initialize_build_system)
  CMakeLists.txt:13 (include)


[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

Note how there is no message actually explaining what the problem in the FPP was.

## Context / Environment

```
Operating System: Linux
CPU Architecture: x86_64
Platform: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.34
Python version: 3.9.18
CMake version: 3.26.5
Pip version: 21.2.3
Pip packages:
    fprime-tools==3.4.5a1
    fprime-gds==3.4.4a3
    fprime-fpp-*==2.2.0a5
```

## How to Reproduce

1. Make a typo in a `.fpp` file, such as `output port array[2]: Exception`
2. Run `fprime-util purge`
3. Run `fprime-util generate --ninja`

## Expected Behavior

Here is the output produced by `fprime-util generate` that should also be produced by `fprime-util generate --ninja`:

```
[INFO] Generating build directory at: [...]
[INFO] Using toolchain file [...]/cmake/toolchain/[...].cmake for platform default
-- The C compiler identification is Clang 18.1.3
-- The CXX compiler identification is Clang 18.1.3
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [python3] python3 found at: /usr/bin/python3
-- [fpp-tools] fpp-depend found at: /usr/local/bin/fpp-depend
-- [fpp-tools] fpp-to-xml found at: /usr/local/bin/fpp-to-xml
-- [fpp-tools] fpp-to-cpp found at: /usr/local/bin/fpp-to-cpp
-- [fpp-tools] fpp-to-dict found at: /usr/local/bin/fpp-to-dict
-- [fpp-tools] fpp-locate-defs found at: /usr/local/bin/fpp-locate-defs
-- [FPRIME] Module locations: [...]
-- [FPRIME] Configuration module: [...]
-- [FPRIME] Installation directory: [...]
-- Target build toolchain/platform: [...]
-- Including [...]/cmake/platform/[...]
-- [sub-build] Performing sub-build: info-cache
-- [sub-build] Generating: info-cache with target/fpp_locs;target/fpp_depend
-- [sub-build] Executing: info-cache with fpp_locs
CMake Error at tools/fprime/cmake/utilities.cmake:586 (message):
  [sub-build] Failed to execute: info-cache/fpp_locs:

  fpp-locate-defs

  [...]/[ComponentName]/[ComponentName].fpp:7.35

              output port array[2]: Exception
                               ^

  error: : expected

  gmake[3]: *** [CMakeFiles/fpp_locs.dir/build.make:74: locs.fpp] Error 1

  gmake[3]: *** Deleting file 'locs.fpp'

  gmake[2]: *** [CMakeFiles/Makefile2:2245: CMakeFiles/fpp_locs.dir/all]
  Error 2

  gmake[1]: *** [CMakeFiles/Makefile2:2252: CMakeFiles/fpp_locs.dir/rule]
  Error 2

  gmake: *** [Makefile:124: fpp_locs] Error 2
Call Stack (most recent call first):
  tools/fprime/cmake/sub-build/sub-build.cmake:65 (execute_process_or_fail)
  tools/fprime/cmake/FPrime.cmake:162 (run_sub_build)
  tools/fprime/cmake/FPrime.cmake:231 (fprime_initialize_build_system)
  CMakeLists.txt:13 (include)


-- Configuring incomplete, errors occurred!
[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

--- a/cmake/utilities.cmake
+++ b/cmake/utilities.cmake
@@ -562,7 +562,7 @@ endfunction(introspect)
 # - **ERROR_MESSAGE**: message to output should an error occurs
 ####
 function(execute_process_or_fail ERROR_MESSAGE)
-    # Quiet standard output unless we are doing verbose output generate
+    # Quiet standard output unless we are doing verbose output (handled below)
     set(OUTPUT_ARGS OUTPUT_QUIET)
     # Print the invocation if debug output is set
     if (CMAKE_DEBUG_OUTPUT)
@@ -571,19 +571,28 @@ function(execute_process_or_fail ERROR_MESSAGE)
         foreach(ARG IN LISTS ARGN)
             set(COMMAND_AS_STRING "${COMMAND_AS_STRING}\"${ARG}\" ")
         endforeach()
-        
-        #string(REPLACE ";" "\" \"" COMMAND_AS_STRING "${ARGN}")
         message(STATUS "[cli] ${COMMAND_AS_STRING}")
+    endif()
+    # Ninja pipes stderr to stdout so remove quiet output to see errors
+    if (CMAKE_GENERATOR MATCHES "Ninja")
+        set(OUTPUT_ARGS)
     endif()
     execute_process(
         COMMAND ${ARGN}
         RESULT_VARIABLE RETURN_CODE
+        OUTPUT_VARIABLE STANDARD_OUTPUT
         ERROR_VARIABLE STANDARD_ERROR
         ERROR_STRIP_TRAILING_WHITESPACE
+        OUTPUT_STRIP_TRAILING_WHITESPACE
         ${OUTPUT_ARGS}
     )
     if (NOT RETURN_CODE EQUAL 0)
-        message(FATAL_ERROR "${ERROR_MESSAGE}:\n${STANDARD_ERROR}")
+        set(MESSAGE "${ERROR_MESSAGE}:\n${STANDARD_ERROR}")
+        # Ninja pipes stderr to stdout so we have to print stdout to see errors
+        if (CMAKE_GENERATOR MATCHES "Ninja")
+            string(APPEND MESSAGE "\n${STANDARD_OUTPUT}")
+        endif()
+        message(FATAL_ERROR "${MESSAGE}")
     endif()
 endfunction()
 

--- a/cmake/utilities.cmake
+++ b/cmake/utilities.cmake
@@ -587,12 +587,12 @@ function(execute_process_or_fail ERROR_MESSAGE)
         ${OUTPUT_ARGS}
     )
     if (NOT RETURN_CODE EQUAL 0)
-        set(MESSAGE "${ERROR_MESSAGE}:\n${STANDARD_ERROR}")
+        set(FATAL_MESSAGE "${ERROR_MESSAGE}:\n${STANDARD_ERROR}")
         # Ninja pipes stderr to stdout so we have to print stdout to see errors
         if (CMAKE_GENERATOR MATCHES "Ninja")
-            string(APPEND MESSAGE "\n${STANDARD_OUTPUT}")
+            string(APPEND FATAL_MESSAGE "\n${STANDARD_OUTPUT}")
         endif()
-        message(FATAL_ERROR "${MESSAGE}")
+        message(FATAL_ERROR "${FATAL_MESSAGE}")
     endif()
 endfunction()
 

